chromium_src/chrome/browser/ui/views/profiles/avatar_toolbar_button_state_manager.cc

```diff
@@ -0,0 +1,118 @@
+// Please see comments on IsRegularProfile() override for more details.
+std::unique_ptr<StateProvider> CreateBraveStateProvider(
+    Profile* profile,
+    StateObserver* state_observer);
+
+}  // namespace
+
+// Overrides profile->IsRegularProfile() in CreateStatesAndListeners() to
+// customize state provider for Tor, Incognito, and Guest profiles.
+// StateProvider provides data to visualize the avatar toolbar button's state
+// according to various situations. But Brave has specific requirements for
+// Tor, Incognito, and Guest profiles to show different icons for these. So
+// in these cases, we use Brave-specific state providers and do not add any
+// other state providers.
+#define IsRegularProfile()                                                 \
+  IsTor() || profile->IsIncognitoProfile() || profile->IsGuestSession()) { \
+    states_[ButtonState::kNormal] =                                        \
+        CreateBraveStateProvider(profile, /*state_observer=*/this);        \
+    return;                                                                \
+  }                                                                        \
+  if (profile->IsRegularProfile()
+
+#include <chrome/browser/ui/views/profiles/avatar_toolbar_button_state_manager.cc>
+
+#undef IsRegularProfile
+
+namespace {
+
+// Brave-specific state providers for the avatar toolbar button. Customizes
+// icons and tooltip texts for Tor, Incognito, and Guest profiles.
+
+// BraveTorStateProvider is used for Tor profiles. As Tor profile is a special
+// case of Brave, we need to provide a icon and tooltip text for this.
+class BraveTorStateProvider : public PrivateBaseStateProvider {
+ public:
+  BraveTorStateProvider(Profile* profile, StateObserver* state_observer)
+      : PrivateBaseStateProvider(profile, state_observer) {}
+
+  // PrivateBaseStateProvider:
+  ui::ImageModel GetAvatarIcon(
+      int icon_size,
+      SkColor icon_color,
+      const ui::ColorProvider& color_provider) const override {
+    return ui::ImageModel::FromVectorIcon(
+        kLeoProductTorIcon, SkColorSetRGB(0x3C, 0x82, 0x3C), icon_size);
+  }
+
+  std::u16string GetText() const override {
+    return l10n_util::GetStringUTF16(IDS_TOR_AVATAR_BUTTON_TOOLTIP_TEXT);
+  }
+};
+
+// BraveIncognitoStateProvider is used for Incognito profiles to override
+// Chromium's default icon for the profile
+class BraveIncognitoStateProvider : public IncognitoStateProvider {
+ public:
+  BraveIncognitoStateProvider(Profile* profile, StateObserver* state_observer)
+      : IncognitoStateProvider(profile, state_observer) {}
+
+  // IncognitoStateProvider:
+  ui::ImageModel GetAvatarIcon(
+      int icon_size,
+      SkColor icon_color,
+      const ui::ColorProvider& color_provider) const override {
+    return ui::ImageModel::FromVectorIcon(
+        kIncognitoIcon, SkColorSetRGB(0xFF, 0xFF, 0xFF), icon_size);
+  }
+};
+
+// BraveGuestStateProvider is used for Guest sessions to override Chromium's
```

Comment:
can't we just override `GetGuestAvatar`?

Reply:
The current override makes them all consistent in terms of overrides, but why do they all use `ButtonState::kNormal`? That seems odd?

Reply:
> can't we just override GetGuestAvatar?
> the current override makes them all consistent in terms of overrides

Yes, I tried to keep these like what upstream is doing.

> why do they all use ButtonState::kNormal

The previous implementation didn't have variation for other states like sync error, sign in, etc when it's guest, tor, or incognito window. I don't think we need to show those cases on those windows? If we do, I think our design team should investigate it first.

Reply:
There should be some explanation in the comments

Reply:
Do you think comment in L22-28 is insufficient?

Reply:
sorry, I'm talking specifically about `ButtonState::kNormal`

Reply:
Appended comment about `ButtonState::Normal`. 8f053d7b0ec1eb0cbdef3abd85b0b6c084edb60e